File: packages/contracts/src/provider-accounts.ts
import { z } from 'zod';
import { AgentKindSchema } from './agent.ts';

/*
 * An account of an agent CLI is one config folder of that CLI, handed to it through the variable it
 * reads its folder from. The person signs in with the CLI itself; the daemon only asks the CLI who
 * that is. Accounts belong to one machine and live in `$RUIMTE_HOME/providers.json`.
 */

/* A slug, never a path. The default account of a CLI has the CLI's kind as its id. */
export const ProviderAccountIdSchema = z.string().regex(/^[a-z][a-z0-9_-]{0,63}$/);
export type ProviderAccountId = z.infer<typeof ProviderAccountIdSchema>;

/*
 * A variable the CLI of an account is started with. A sensitive value lives in the keychain of the
 * machine and never on the wire: the daemon sends it as an empty `value` with `valueRedacted`, and a
 * save that sends that back keeps what the keychain holds.
 */
export const ProviderAccountVariableSchema = z.object({
    name: z.string().regex(/^[A-Za-z_][A-Za-z0-9_]{0,127}$/),
    value: z.string(),
    sensitive: z.boolean(),
    valueRedacted: z.boolean().optional()
});
export type ProviderAccountVariable = z.infer<typeof ProviderAccountVariableSchema>;

export const ProviderAccountSchema = z.object({
    // An agent kind. A kind this version does not know still parses, so the entry is written back unchanged.
    kind: z.string().min(1),
    label: z.string().optional(),
    // A node accent name (`NODE_ACCENT_NAMES`), kept a plain string so an older name is written back; the client paints any other with its accent.
    color: z.string().optional(),
    // Absent is on. An account that is off is left out of every picker.
    enabled: z.boolean().optional(),
    // The CLI's config folder, `~` allowed. Absent on the default account, which uses the CLI's own folder.
    home: z.string().min(1).optional(),
    // Codex only: a folder of its own for the login, sharing everything else with `home`.
    shadowHome: z.string().min(1).optional(),
    // Set on the CLI's environment after the login variables are taken off, so a key here is one a person chose.
    env: z.array(ProviderAccountVariableSchema).optional()
});
export type ProviderAccount = z.infer<typeof ProviderAccountSchema>;

export const ProviderAccountStateSchema = z.enum([
    'checking',
    'disabled',
    // The CLI is not installed on this machine.
    'not-found',
    'signed-out',
    'ready',
    'folder-missing',
    // A kind this version does not know, or a CLI with no variable for its config folder.
    'unavailable',
    'failed'
]);
export type ProviderAccountState = z.infer<typeof ProviderAccountStateSchema>;

/* What the CLI said about an account the last time it was asked. Never stored: the CLI is the truth. */
export const ProviderAccountStatusSchema = z.object({
    id: ProviderAccountIdSchema,
    kind: z.string(),
    state: ProviderAccountStateSchema,
    email: z.string().nullable(),
    plan: z.string().nullable(),
    organization: z.string().nullable(),
    // The folder the CLI is started with, resolved; empty for a CLI without a config folder variable.
    home: z.string(),
    // Where the account's conversations are written, resolved: a chat goes on under another account only when both write here.
    transcripts: z.string().optional(),
    // Why the state is what it is, in a sentence, when the CLI or the daemon said.
    message: z.string().nullable(),
    // Milliseconds since the epoch; 0 while the account was never checked.
    checkedAt: z.number()
});
export type ProviderAccountStatus = z.infer<typeof ProviderAccountStatusSchema>;

export const ProviderAccountMapSchema = z.record(ProviderAccountIdSchema, ProviderAccountSchema);
export type ProviderAccountMap = z.infer<typeof ProviderAccountMapSchema>;

export const ProviderAccountsSchema = z.object({
    accounts: ProviderAccountMapSchema,
    statuses: z.array(ProviderAccountStatusSchema),
    // Whether this machine can keep a sensitive variable; false off macOS, where no keychain is used.
    secretsAvailable: z.boolean().optional(),
    // Per agent kind, what a terminal types to sign that CLI in under an account, which the account's environment points at.
    loginCommands: z.partialRecord(AgentKindSchema, z.string()).optional()
});
export type ProviderAccounts = z.infer<typeof ProviderAccountsSchema>;

/* The whole map: an id left out is removed, and its folder stays where it is. */
export const ProviderAccountsSavePayloadSchema = z.object({
    accounts: ProviderAccountMapSchema
});
export type ProviderAccountsSavePayload = z.infer<typeof ProviderAccountsSavePayloadSchema>;

/* An account in a folder the daemon makes for it under `$RUIMTE_HOME/accounts`; a Codex one is a shadow home over the CLI's own folder. */
export const ProviderAccountCreatePayloadSchema = z.object({
    kind: AgentKindSchema,
    label: z.string().trim().min(1).max(80),
    color: z.string().optional()
});
export type ProviderAccountCreatePayload = z.infer<typeof ProviderAccountCreatePayloadSchema>;

export const ProviderAccountCreateResultSchema = ProviderAccountsSchema.extend({
    // The id the daemon minted from the label.
    id: ProviderAccountIdSchema
});
export type ProviderAccountCreateResult = z.infer<typeof ProviderAccountCreateResultSchema>;

/* Asks the CLI of one account again every few seconds until it is signed in, for a login running in a terminal. */
export const ProviderAccountWatchLoginPayloadSchema = z.object({
    id: ProviderAccountIdSchema
});
export type ProviderAccountWatchLoginPayload = z.infer<typeof ProviderAccountWatchLoginPayloadSchema>;

File: apps/ios/Packages/RuimtePulsar/Sources/RuimtePulsar/ProviderAccounts.swift
import Foundation

/// One account of an agent CLI on a machine, as `accounts.list` describes it. The default account of a CLI has the
/// CLI's kind as its id.
public struct ProviderAccountEntry: Sendable, Equatable, Identifiable {
    public let id: String
    public let kind: String
    public let label: String?
    /// A node accent id; any other value paints with the app's accent.
    public let color: String?
    public let enabled: Bool
    /// `ProviderAccountState` as the machine last saw it; nil before the machine described the account.
    public let state: String?
    /// Where the account's conversations are written; nil when the machine does not say.
    public let transcripts: String?
    public let message: String?

    public init(
        id: String, kind: String, label: String? = nil, color: String? = nil, enabled: Bool = true,
        state: String? = nil, transcripts: String? = nil, message: String? = nil
    ) {
        self.id = id
        self.kind = kind
        self.label = label
        self.color = color
        self.enabled = enabled
        self.state = state
        self.transcripts = transcripts
        self.message = message
    }

    public var isDefault: Bool { id == kind }

    /// A default account nobody named goes by its CLI's name.
    public func name(provider: String) -> String { label ?? (isDefault ? provider : id) }
}

/// The accounts of every CLI on one machine.
public struct ProviderAccountList: Sendable, Equatable {
    public let entries: [ProviderAccountEntry]

    public init(entries: [ProviderAccountEntry]) {
        self.entries = entries
    }

    /// Reads a `accounts.list` result or a `accounts.changed` payload. The account map arrives as an object, which
    /// keeps no order, so the order is the machine's list of statuses, and an account without one comes last by id.
    public init(_ value: JSONValue) {
        let statuses = value["statuses"]?.arrayValue ?? []
        let order = Dictionary(
            statuses.enumerated().compactMap { index, status in status["id"]?.stringValue.map { ($0, index) } },
            uniquingKeysWith: { first, _ in first })
        let statusByID = Dictionary(
            statuses.compactMap { status in status["id"]?.stringValue.map { ($0, status) } },
            uniquingKeysWith: { first, _ in first })
        let accounts = (value["accounts"]?.objectValue ?? [:]).sorted { first, second in
            switch (order[first.key], order[second.key]) {
            case (let one?, let other?): return one < other
            case (.some, nil): return true
            case (nil, .some): return false
            case (nil, nil): return first.key < second.key
            }
        }
        entries = accounts.compactMap { id, account in
            guard let kind = account["kind"]?.stringValue else { return nil }
            let status = statusByID[id]
            return ProviderAccountEntry(
                id: id, kind: kind, label: account["label"]?.stringValue, color: account["color"]?.stringValue,
                enabled: account["enabled"]?.boolValue ?? true, state: status?["state"]?.stringValue,
                transcripts: status?["transcripts"]?.stringValue, message: status?["message"]?.stringValue)
        }
    }

    /// The accounts of one CLI, its default one first.
    public func accounts(of kind: String) -> [ProviderAccountEntry] {
        let own = entries.filter { $0.kind == kind }
        return own.filter(\.isDefault) + own.filter { !$0.isDefault }
    }

    /// Whether a person has a choice of account for this CLI, which is when a chat names its account.
    public func hasChoice(_ kind: String) -> Bool { accounts(of: kind).filter(\.enabled).count >= 2 }

    /// The accounts a picker offers: every one that is on, and the one in use even while it is off.
    public func offered(_ kind: String, current: String?) -> [ProviderAccountEntry] {
        let current = current ?? kind
        return accounts(of: kind).filter { $0.enabled || $0.id == current }
    }

    /// Whether a chat of one account can go on under another, the way the machine decides it: the same account, or two
    /// of the chat's CLI that write their conversations to one folder. A machine that does not say where answers no,
    /// so the person forks instead of being refused.
    public func canContinue(_ kind: String, from: String?, to: String?) -> Bool {
        let fromID = from ?? kind
        let toID = to ?? kind
        if fromID == toID { return true }
        let folder = { (id: String) in entries.first { $0.id == id && $0.kind == kind }?.transcripts }
        guard let source = folder(fromID), !source.isEmpty else { return false }
        return source == folder(toID)
    }

    /// Whether a chat has spoken, after which only an account that reads its conversation can take it over.
    public static func chatStarted(_ info: JSONValue) -> Bool {
        let session = info["agentSessionId"]
        return (session != nil && session != .null) || (info["usage"]?["turns"]?.numberValue ?? 0) > 0
    }
}
